src/EditorFeatures/TestUtilities/Threading/WpfTestRunner.cs

```diff
@@ -50,39 +50,73 @@ public sealed class WpfTestRunner : XunitTestRunner
         protected override Task<decimal> InvokeTestMethodAsync(ExceptionAggregator aggregator)
         {
             SharedData.ExecutingTest(TestMethod);
-            var sta = StaTaskScheduler.DefaultSta;
+
+            DispatcherSynchronizationContext synchronizationContext = null;
+            Dispatcher dispatcher = null;
+            using (var staThreadStartedEvent = new ManualResetEventSlim(initialState: false))
+            {
+                var staThread = new Thread((ThreadStart)(() =>
+                {
+                    // All WPF Tests need a DispatcherSynchronizationContext and we dont want to block pending keyboard
+                    // or mouse input from the user. So use background priority which is a single level below user input.
+                    synchronizationContext = new DispatcherSynchronizationContext();
+                    dispatcher = Dispatcher.CurrentDispatcher;
+
+                    // xUnit creates its own synchronization context and wraps any existing context so that messages are
+                    // still pumped as necessary. So we are safe setting it here, where we are not safe setting it in test.
+                    SynchronizationContext.SetSynchronizationContext(synchronizationContext);
+
+                    staThreadStartedEvent.Set();
+
+                    Dispatcher.Run();
+                }));
+
+                staThread.Name = nameof(WpfTestRunner);
+                staThread.SetApartmentState(ApartmentState.STA);
+                staThread.Start();
+
+                staThreadStartedEvent.Wait();
+                Debug.Assert(synchronizationContext != null);
+            }
+
+            var taskScheduler = new SynchronizationContextTaskScheduler(synchronizationContext);
             var task = Task.Factory.StartNew(async () =>
             {
-                Debug.Assert(sta.StaThread == Thread.CurrentThread);
+                Debug.Assert(SynchronizationContext.Current is DispatcherSynchronizationContext);
 
                 using (await SharedData.TestSerializationGate.DisposableWaitAsync(CancellationToken.None))
                 {
-                    try
-                    {
-                        Debug.Assert(SynchronizationContext.Current is DispatcherSynchronizationContext);
+                    // Sync up FTAO to the context that we are creating here. 
+                    ForegroundThreadAffinitizedObject.CurrentForegroundThreadData = new ForegroundThreadData(
+                        Thread.CurrentThread,
+                        new SynchronizationContextTaskScheduler(new DispatcherSynchronizationContext(Dispatcher.CurrentDispatcher, DispatcherPriority.Background)),
```

Comment:
Technically `taskScheduler` uses `DispatcherPriority.Normal` where the one here uses `DispatcherPriority.Background`. It's a bit concerning since `taskScheduler` is *documented* to use "background priority". This isn't a behavior change from the current state of the target branch so I'll probably leave it for now, but it's concerning enough to file a follow-up issue (I'll file the issue after merging if it's not corrected before then).